staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go

```diff
@@ -288,33 +298,74 @@ type Backoff struct {
 	Cap time.Duration
 }
 
-// Step (1) returns an amount of time to sleep determined by the
-// original Duration and Jitter and (2) mutates the provided Backoff
-// to update its Steps and Duration.
+// Step returns an amount of time to sleep determined by the original
+// Duration and Jitter. The backoff is mutated to update its Steps and
+// Duration.
 func (b *Backoff) Step() time.Duration {
-	if b.Steps < 1 {
-		if b.Jitter > 0 {
-			return Jitter(b.Duration, b.Jitter)
-		}
-		return b.Duration
-	}
-	b.Steps--
+	var nextDuration time.Duration
+	nextDuration, b.Duration, b.Steps = delay(b.Steps, b.Duration, b.Cap, b.Factor, b.Jitter)
+	return nextDuration
+}
 
+// delayFunc returns a function that will compute the next interval to
+// wait given the arguments in b. It does not mutate the original backoff
+// but the function is safe to use only from a single goroutine.
+func (b Backoff) delayFunc() DelayFunc {
```

Comment:
we should choose value or pointer receiver for Backoff, it seems value is ok since we don't modify the type,  but the `b.Step()` method in line294 is defined as receiver

Reply:
I defer to someone with more knowledge on the golang internals, I just flagged  it because I couldn't find if is correct or no,  only found this reference  https://go.dev/tour/methods/8

> In general, all methods on a given type should have either value or pointer receivers, but not a mixture of both. (We'll see why over the next few pages.)

Reply:
Reason to have the separation

1. Existing method needs pointer receiver to work (backward compat for all existing code, no breakages)
2. New methods are non-mutating and deliberately designed to be easily inlined: `Backoff{Step: 3, Duration: time.Millisecond}.Timer()`.  If we required a pointer, all inline call sites get uglier `(&Backoff{Step: 3, Duration: time.Millisecond}).Timer()` which is more work for the user and more confusing.

Sufficiently justified?